selectivecar.go

```diff
@@ -0,0 +1,251 @@
+package car
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+
+	util "github.com/ipfs/go-car/util"
+	cid "github.com/ipfs/go-cid"
+	"github.com/ipld/go-ipld-prime"
+	dagpb "github.com/ipld/go-ipld-prime-proto"
+	ipldfree "github.com/ipld/go-ipld-prime/impl/free"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	"github.com/ipld/go-ipld-prime/traversal"
+	"github.com/ipld/go-ipld-prime/traversal/selector"
+)
+
+// Dag is a root/selector combo to put into a car
+type Dag struct {
+	Root     cid.Cid
+	Selector ipld.Node
+}
+
+// Block is all information and metadata about a block that is part of a car file
+type Block struct {
+	BlockCID cid.Cid
+	Data     []byte
+	Offset   uint64
+	Size     uint64
+}
+
+// SelectiveCar is a car file based on root + selector combos instead of just
+// a single root and complete dag walk
+type SelectiveCar struct {
+	ctx   context.Context
+	dags  []Dag
+	store ReadStore
+}
+
+// OnCarHeaderFunc is called during traversal when the header is created
+type OnCarHeaderFunc func(CarHeader) error
+
+// OnNewCarBlockFunc is called during traveral when a new unique block is encountered
+type OnNewCarBlockFunc func(Block) error
+
+// SelectiveCarPrepared is a SelectiveCar that has already been traversed, such that it
+// can be written quicker with Dump. It also contains metadata already collection about
+// the Car file like size and number of blocks that go into it
+type SelectiveCarPrepared struct {
+	SelectiveCar
+	size   uint64
+	header CarHeader
+	cids   []cid.Cid
+}
+
+// NewSelectiveCar creates a new SelectiveCar for the given car file based
+// a block store and set of root+selector pairs
+func NewSelectiveCar(ctx context.Context, store ReadStore, dags []Dag) SelectiveCar {
```

Comment:
@mikeal considering the next iteration of CAR, should we start making new APIs only take a single root rather than allowing for an array of roots? I can't recall exactly now but I think we were all agreeing that we should make it strictly one root and make the user work around that.

Reply:
Note that go-ipfs will need to continue accepting multiple roots / no-roots / incomplete DAGs due to non-fil specific use-cases. I.e. keep the door open to "sloppy" or "strict=false" interfaces when making modifications.

Reply:
(a) let's enumerate those use-cases because it informs CARv2
(b) this is only for write interfaces, and afaik your exporter is producing CAR files with roots - can they be more than one root?

Let's get CARv2 discussions going before you ship go-ipfs with CAR import/export because I don't want us to lock in CAR to supporting more than we want it to support into the future.

Reply:
For the foreseable future go-ipfs will *export* ( produce ) .car files with a single root, with the entire requisite DAG being present in the produced file. I.e. the only cars go-ipfs will be producing at lest in v0.1 are "fil-compliant .car's"

However the .car *import* functionality is nothing more than a generalization of `ipfs block put`, unblocking a plethora of use-cases currently relying on the inefficient one-at-a-time interface.

In other words: if go-car ( the library ) starts validating a .car file before "streaming it out" - there needs to be a way for the caller to disable these checks, because go-ipfs must be able to import:
- a .car with disjoint blocks and no roots
- a dag split over multiple .car files with the root specified only in one of the headers
- a mixture of the above with multiple roots specified in multiple headers
- roots specified in headers *without* the entire dag being contained  in *any* of the supplied .car files ( supplementary / update )

Let me know if this is not sufficient description and we could jump on a call / get a more proper doc going.

Reply:
I would prefer not to try to incorporate things that anticipate the next version of CAR until it settles, and then adjust then.

Reply:
yeah ok, so I chatted with @mikeal about this and he still sees the multi-root use-case sticking around and that'll probably end up in CARv2, so we can move this discussion to ipld/specs when I get around (soon) to doing some initial CARv2 work.